Introduce React Testing Library

labkey-nicka opened the pull request:
#### Rationale
We recently discovered that [Enzyme](https://enzymejs.github.io/enzyme/), our JavaScript testing library, [is and will not be compatible with React v18+](https://stackoverflow.com/a/73999396). This PR introduces [React Testing Library](https://testing-library.com/docs/react-testing-library/intro/) (RTL) as a second supported test library within `@labkey/components`. The intent here is to be able to start writing new tests using RTL and to allow for the team to familiarize themselves with the new framework. Soon™ after that we can stop writing any new Enzyme tests and start on migrating existing tests to RTL. Here is how I've delineated these frameworks:

**Enzyme**
- Command: `yarn test-enzyme`
- File suffix: `.spec.ts(x)`
- Utilities: `enzymeTestHelpers.tsx`

**React Testing Library**
- Command: `yarn test-react`
- File suffix: `.test.ts(x)`
- Utilities: `reactTestLibraryHelpers.tsx`

All tests can be run via `yarn test`. I'm working with @labkey-tchad to configure CI to correctly run and report these tests.

**Useful Resources**
- [React Testing Library Docs](https://testing-library.com/docs/react-testing-library/intro)
- [Migrate from Enzyme](https://testing-library.com/docs/react-testing-library/migrate-from-enzyme)
- [Common mistakes with React Testing Library](https://kentcdodds.com/blog/common-mistakes-with-react-testing-library)

#### Related Pull Requests
- https://github.com/LabKey/labkey-ui-components/pull/1206
- https://github.com/LabKey/labkey-ui-premium/pull/107
- https://github.com/LabKey/biologics/pull/2170
- https://github.com/LabKey/sampleManagement/pull/1880
- https://github.com/LabKey/inventory/pull/882

#### Changes
- Update `@labkey/build` to skip processing of `**/*.*test.*` files and declare `@testing-library` as externals.
- Add `test-enzyme` and `test-react` scripts. Call both via `yarn test`.
- Declare alternative `jest.setup.react.ts` to support RTL-specific imports.
- Declare `enzymeTestHelpers` and `reactTestLibraryHelpers` to house utilities specific to each library.
- Add initial example tests for `LabelHelpTip`, `BooleanFieldOptions` and `DesignerDetailPanel`.

labkey-nicka committed:
Add @testing-library/react dependency

labkey-nicka committed:
react-testing-library configuration

labkey-nicka committed:
Example test: LabelHelpTip.test.tsx

labkey-nicka committed:
nit: simplify callback on LabelHelpTip

labkey-nicka committed:
testUtils: renderWithAppContext

labkey-nicka committed:
DesignerDetailPanel.test.tsx

labkey-nicka committed:
enzymeTestHelpers, reactTestLibraryHelpers

labkey-nicka committed:
Export renderWithAppContext

labkey-nicka committed:
Build updates for react-testing-library

labkey-nicka committed:
@labkey/build@6.11.0-fb-react-test-lib.0

labkey-nicka committed:
Bump @labkey/build

labkey-nicka committed:
BooleanFieldOptions

labkey-nicka committed:
Update import paths

labkey-nicka committed:
Update test snapshots

labkey-nicka committed:
Remove unused import

labkey-nicka committed:
AdminSettingsPage: update test to compile

labkey-nicka committed:
Rename jest.config.js -> jest.enzyme.config.js

labkey-nicka committed:
Update test commands

labkey-nicka committed:
Remove comment

cnathe commented on packages/components/package.json:
```diff
@@ -25,8 +25,10 @@
-    "test-ci": "cross-env NODE_ENV=test jest --ci --silent",
+    "test": "npm run test-react ; npm run test-enzyme",
+    "test-enzyme": "cross-env NODE_ENV=test jest --config jest.enzyme.config.js --maxWorkers=6 --silent",
+    "test-react": "cross-env NODE_ENV=test jest --config jest.react.config.js",
```
should either of the following flags be set for `test-react` as well? `--maxWorkers=6 --silent`

labkey-nicka replied:
I've added `--maxWorkers=6` so it doesn't over consume processors on dev machines. I've left `--silent` off in the hopes that we can squash failures/warnings for these from the outset.

cnathe commented on packages/components/src/internal/components/base/LabelHelpTip.test.tsx:
```diff
@@ -0,0 +1,48 @@
+        fireEvent.mouseOver(wrapper);
+
+        const contentElement = document.querySelector('.test-content');
+        expect(contentElement).toBeVisible();
```
is there a difference between `toBeInTheDocument()` and `toBeVisible()`?

labkey-nicka replied:
`toBeInTheDocument()` ([docs](https://github.com/testing-library/jest-dom/#tobeinthedocument)) asserts whether an element is present in the document or not. `toBeVisible()` ([docs](https://github.com/testing-library/jest-dom/#tobevisible)) checks if an element is currently visible to the user (given necessary documented conditions for what "visible" means).

labkey-tchad commented on packages/components/package.json:
```diff
@@ -25,8 +25,10 @@
+    "test": "npm run test-react ; npm run test-enzyme",
+    "test-enzyme": "cross-env NODE_ENV=test jest --config jest.enzyme.config.js --maxWorkers=6 --silent",
+    "test-react": "cross-env NODE_ENV=test jest --config jest.react.config.js",
+    "test-ci": "cross-env NODE_ENV=test jest --config jest.enzyme.config.js --ci --silent",
```
Maybe make `test-ci` just run them both so TeamCity doesn't have to worry about the different test types?
```suggestion
    "test-ci": "npm run test-react ; cross-env NODE_ENV=test jest --config jest.enzyme.config.js --ci --silent",
```

labkey-nicka replied:
I didn't make exactly the suggested change but this appears to work well in reporting in CI!

labkey-nicka committed:
Add maxWorkers to test-react

labkey-nicka committed:
Try integrated test-ci command

labkey-nicka committed:
Migrate to testHelpers

labkey-nicka committed:
Document renderWithAppContext

labkey-nicka committed:
Move enzymeTestHelpers.tsx

labkey-nicka committed:
Revert jest.enzyme.config.js -> jest.config.js
- Renaming away from the default name for this file breaks IntelliJ's built-in configuration matching

labkey-alan commented on packages/components/jest.config.js:
Should we consider naming this file `jest.enzyme.config.js`?

labkey-nicka replied:
Heh, so I did that but then realized that it makes it so IntelliJ doesn't recognize how to run the unit tests directly since it assumes the presence of a `jest.config.js`. There may be some way to do this but for now I've reverted this and we can switch these around when we want to move/default to RTL.

<img width="826" alt="image" src="https://github.com/LabKey/labkey-ui-components/assets/3926239/3f8d71d5-d36e-48b1-be21-db30620c2a10">

labkey-alan reviewed (approved):
Looks good to me, exciting to see this test framework looks a lot more like using the regular browser APIs.

labkey-nicka committed:
Remove package export of renderWithAppContext
- causes errors in downstream packages needing to resolve @testing-library/react
- will need to resolve if test utilities are to be shared across packages

labkey-alan commented on packages/components/src/internal/components/domainproperties/BooleanFieldOptions.test.tsx:
```diff
@@ -0,0 +1,63 @@
+
+        const elementId = createFormInputId(DOMAIN_FIELD_FORMAT, domainIndex, index);
+        const formatField = document.getElementById(elementId);
+        fireEvent.change(formatField, { target: { value: expectedValue } });
```
Should we consider [following the advice](https://kentcdodds.com/blog/common-mistakes-with-react-testing-library#not-using-testing-libraryuser-event) from the "Common Mistakes" article, and use `@testing-library/user-event`?

labkey-nicka replied:
Added.

labkey-nicka committed:
Add package export of renderWithAppContext

labkey-nicka committed:
Make @testing-library a dependency

labkey-nicka committed:
Add @testing-library/user-event

labkey-nicka committed:
Add @testing-library/user-event to package externals

labkey-nicka committed:
@labkey/build@6.11.0-fb-react-test-lib.1

labkey-nicka committed:
Use userEvent

labkey-nicka committed:
Bump @labkey/build

labkey-nicka committed:
v2.343.1-fb-react-test-lib.0

labkey-nicka committed:
Prepare release notes

labkey-nicka committed:
@labkey/build@6.12.0

labkey-nicka committed:
Bump @labkey/build

labkey-nicka committed:
Prepare release notes

labkey-nicka committed:
lint

labkey-nicka committed:
v2.344.0
